### ISSUE TITLE
Crash when opening a zone with inconsistent comments in the same records.

### ISSUE THREAD

**Issue by brlumen** [ngoduykhanh/PowerDNS-Admin#720]
image: ngoduykhanh/powerdns-admin:latest
Database provider: mariadb  Ver 15.1 Distrib 10.4.12-MariaDB, for Linux (x86_64) using readline 5.1
Pdns:  PowerDNS Authoritative Server 4.3.0

Error reproducing method:

1. Create an A-record "test" with ip 127.0.0.1 and any comment.
2. Create an A-record "test" with ip 127.0.0.2 without comment.
3. Reload the page.

The following message appears in the container log:
```powerdns_admin | [2020-05-07 10:28:50 +0000] [11] [DEBUG] GET /domain/example.com
powerdns_admin | [2020-05-07 10:28:50 +0000] [11] [DEBUG] Fetched rrests:
powerdns_admin | [
powerdns_admin |     {
powerdns_admin |         "comments": [
powerdns_admin |             {
powerdns_admin |                 "account": "",
powerdns_admin |                 "content": "\u0432\u0430\u044b\u0430",
powerdns_admin |                 "modified_at": 1588847328
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "name": "mail.example.com.",
powerdns_admin |         "records": [
powerdns_admin |             {
powerdns_admin |                 "content": "127.0.0.1",
powerdns_admin |                 "disabled": false
powerdns_admin |             },
powerdns_admin |             {
powerdns_admin |                 "content": "127.0.0.2",
powerdns_admin |                 "disabled": false
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "ttl": 3600,
powerdns_admin |         "type": "A"
powerdns_admin |     },
powerdns_admin |     {
powerdns_admin |         "comments": [],
powerdns_admin |         "name": "web.example.com.",
powerdns_admin |         "records": [
powerdns_admin |             {
powerdns_admin |                 "content": "A \"view({{ {'172.27.113.0/24'}, {'192.168.1.54'}},{ {'0.0.0.0/0'}, {'192.0.2.1'}} }) \"",
powerdns_admin |                 "disabled": false
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "ttl": 60,
powerdns_admin |         "type": "LUA"
powerdns_admin |     },
powerdns_admin |     {
powerdns_admin |         "comments": [],
powerdns_admin |         "name": "example.com.",
powerdns_admin |         "records": [
powerdns_admin |             {
powerdns_admin |                 "content": "dns-test.example.com. admin.example.com. 2020050713 10800 3600 604800 3600",
powerdns_admin |                 "disabled": false
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "ttl": 3600,
powerdns_admin |         "type": "SOA"
powerdns_admin |     }
powerdns_admin | ]
powerdns_admin | [2020-05-07 10:28:50,915] [domain.py:61] DEBUG - Fetched rrests:
powerdns_admin | [
powerdns_admin |     {
powerdns_admin |         "comments": [
powerdns_admin |             {
powerdns_admin |                 "account": "",
powerdns_admin |                 "content": "\u0432\u0430\u044b\u0430",
powerdns_admin |                 "modified_at": 1588847328
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "name": "mail.example.com.",
powerdns_admin |         "records": [
powerdns_admin |             {
powerdns_admin |                 "content": "127.0.0.1",
powerdns_admin |                 "disabled": false
powerdns_admin |             },
powerdns_admin |             {
powerdns_admin |                 "content": "127.0.0.2",
powerdns_admin |                 "disabled": false
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "ttl": 3600,
powerdns_admin |         "type": "A"
powerdns_admin |     },
powerdns_admin |     {
powerdns_admin |         "comments": [],
powerdns_admin |         "name": "web.example.com.",
powerdns_admin |         "records": [
powerdns_admin |             {
powerdns_admin |                 "content": "A \"view({{ {'172.27.113.0/24'}, {'192.168.1.54'}},{ {'0.0.0.0/0'}, {'192.0.2.1'}} }) \"",
powerdns_admin |                 "disabled": false
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "ttl": 60,
powerdns_admin |         "type": "LUA"
powerdns_admin |     },
powerdns_admin |     {
powerdns_admin |         "comments": [],
powerdns_admin |         "name": "example.com.",
powerdns_admin |         "records": [
powerdns_admin |             {
powerdns_admin |                 "content": "dns-test.example.com. admin.example.com. 2020050713 10800 3600 604800 3600",
powerdns_admin |                 "disabled": false
powerdns_admin |             }
powerdns_admin |         ],
powerdns_admin |         "ttl": 3600,
powerdns_admin |         "type": "SOA"
powerdns_admin |     }
powerdns_admin | ]
powerdns_admin | [2020-05-07 10:28:50 +0000] [11] [ERROR] Exception on /domain/example.com [GET]
powerdns_admin | Traceback (most recent call last):
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 2446, in wsgi_app
powerdns_admin |     response = self.full_dispatch_request()
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1951, in full_dispatch_request
powerdns_admin |     rv = self.handle_user_exception(e)
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1820, in handle_user_exception
powerdns_admin |     reraise(exc_type, exc_value, tb)
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
powerdns_admin |     raise value
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1949, in full_dispatch_request
powerdns_admin |     rv = self.dispatch_request()
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1935, in dispatch_request
powerdns_admin |     return self.view_functions[rule.endpoint](**req.view_args)
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask_login/utils.py", line 261, in decorated_view
powerdns_admin |     return func(*args, **kwargs)
powerdns_admin |   File "/app/powerdnsadmin/decorators.py", line 60, in decorated_function
powerdns_admin |     return f(*args, **kwargs)
powerdns_admin |   File "/app/powerdnsadmin/routes/domain.py", line 108, in domain
powerdns_admin |     comment=r['comments'][index]['content']
powerdns_admin | IndexError: list index out of range
powerdns_admin | [2020-05-07 10:28:50,953] [app.py:1890] ERROR - Exception on /domain/example.com [GET]
powerdns_admin | Traceback (most recent call last):
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 2446, in wsgi_app
powerdns_admin |     response = self.full_dispatch_request()
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1951, in full_dispatch_request
powerdns_admin |     rv = self.handle_user_exception(e)
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1820, in handle_user_exception
powerdns_admin |     reraise(exc_type, exc_value, tb)
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
powerdns_admin |     raise value
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1949, in full_dispatch_request
powerdns_admin |     rv = self.dispatch_request()
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1935, in dispatch_request
powerdns_admin |     return self.view_functions[rule.endpoint](**req.view_args)
powerdns_admin |   File "/usr/lib/python3.8/site-packages/flask_login/utils.py", line 261, in decorated_view
powerdns_admin |     return func(*args, **kwargs)
powerdns_admin |   File "/app/powerdnsadmin/decorators.py", line 60, in decorated_function
powerdns_admin |     return f(*args, **kwargs)
powerdns_admin |   File "/app/powerdnsadmin/routes/domain.py", line 108, in domain
powerdns_admin |     comment=r['comments'][index]['content']
powerdns_admin | IndexError: list index out of range
```

**Comment by brlumen** [comment]
![image](https://user-images.githubusercontent.com/26021974/81286878-fdbb8400-908b-11ea-9556-b3321a4733c9.png)
And some kind of problem with the order of comments. The number in the comment when adding the entry coincided with the last digit in ip.

**Comment by terbolous** [comment]
Verified. The index out of bounds can be fixed with a patch like this:
```
--- a/powerdnsadmin/routes/domain.py
+++ b/powerdnsadmin/routes/domain.py
@@ -106,7 +106,7 @@ def domain(domain_name):
                         ttl=r['ttl'],
                         data=record['content'],
                         comment=r['comments'][index]['content']
-                        if r['comments'] else '',
+                        if (r['comments'] and len(r['comments']) > index) else '',
                         is_allowed_edit=True)
                     index += 1
                     records.append(record_entry)
```

The last one comes from PowerDNS, there's no correlation between the comments list and record list and thus no easy way to know which record a comment belongs to. Have to check if this has been fixed in newer versions of PowerDNS

**Comment by terbolous** [comment]
For the order error i believe this stems from the handling of record merging before pushing to PDNS. It should be changed so that if a similar record name has a comment an empty one is added to keep the ordering

**Comment by sshipway** [comment]
This explains a few things.  Glad to see there's already a PR to get around it though a full fix would be nice

**Comment by ngoduykhanh** [comment]
Should be fixed since #739 got merged. Feel free to reopen the issue if you face the same problem.
